Export grab bag; history, busy state, sorting, auto-asset creation...

mcilroyc opened the pull request:
A handful of export-related changes...

1. History store now tracks the export store' state, and modifications to export assets create history states.
1. Pursuant to the above, DataList and TextInput were modified so that selecting the same (default) value will not call its `onChange` (thus not causing frivolous history states when tabbing through export configuration values)
1. Artboards in the Export dialog are now sorted via their "grid" layout on canvas as opposed to by Z-index in the layer's panel (this partially mitigates #2212, I think)
1. Artboards in the Export dialog: optionally prefix the filename with a three digit value, generated on the fly by the grid-sorted order of artboards
1. The above "prefix" option is defaulted to false, but the user's choice is sticky, via DS preferences.
1. Export Panel: :arrow_down: icon will auto-generate a default asset if none exist, making simple export a one-click operation
1. Creating a new artboard will auto-populate a simple 1x asset configuration.  Note that it will not be fully supported until the artboard has at least one visible component (otherwise generator will not export it, so it is marked as not export-supported initially)
1. Exports are now done without blocking the UI, and the "busy" state is reflected in both Export dialog and panel.  This UI will need to improve, but this is a start.
1. The folder prompt process will throw a custom error if the user cancels, which allowed for a more readable (IMO) refactoring of the export* actions.
1. Document-level exports will use the document's name as the base file name
1. Addresses disabled state export button icon bug #2302
1. Addresses non-empty-group support bug #2376

mcilroyc commented:
I will tidy up the commits prior to merging, but left it in for granularity of review.

pineapplespatula commented on src/js/actions/export.js:
```diff
@@ -158,7 +182,7 @@ define(function (require, exports) {
                     status: ExportAsset.STATUS.STABLE
                 };
-                return this.transfer(updateExportAsset, document, _layers, assetIndex, assetProps);
+                return this.flux.actions.export.updateExportAsset(document, _layers, assetIndex, assetProps, true);
```
why are we not transferring here?

mcilroyc replied:
see comment at line 158

pineapplespatula replied:
the commend does not say why.

mcilroyc replied:
I've elaborated in the comment, see next commit.

pineapplespatula commented on src/js/actions/export.js:
```diff
@@ -167,11 +191,33 @@ define(function (require, exports) {
+     * @param {Immutable.Iterable.<Promise>} exportList
+     */
+    var _batchExports = function (exportList) {
+        if (_activeExports) {
```
where does _activeExports get set to a non null value?

mcilroyc replied:
4 lines below this one

pineapplespatula commented on src/js/actions/export.js:
```diff
@@ -440,6 +501,34 @@ define(function (require, exports) {
+        if (!layerExports || layerExports.isEmpty()) {
+            return this.transfer(updateExportAsset, document, Immutable.List.of(layer), 0, {}, true);
+        } else {
+            Promise.resolve();
```
should this return here?

mcilroyc replied:
yep, good catch

pineapplespatula commented on src/js/util/gridsort.js:
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2014 Adobe Systems Incorporated. All rights reserved.
```
2015

pineapplespatula commented on src/js/actions/export.js:
```diff
@@ -137,19 +153,27 @@ define(function (require, exports) {
      * @param {number} assetIndex index of this asset within the layer's list
      * @param {ExportAsset} asset asset to export
+     * @param {string=} baseDir Optional directory path in to which assets should be exported
+     * @param {string=} prefix
```
Optional? description?

pineapplespatula commented on src/js/util/gridsort.js:
```diff
@@ -0,0 +1,120 @@
+     * @private
+     * @param {Array.<object>} rows
+     * @param {Layer} layer
+     * @param {number=} index if not supplied, start at zero
```
you usually say "Optional" i'm not sure its needed, but i noticed given the "="  in the type but i thought i would mention it

pineapplespatula commented on src/js/util/gridsort.js:
```diff
@@ -0,0 +1,120 @@
+        if (!row) {
+            newIndex = index;
+        } else if (bounds.bottom < row.top) {
+            newIndex = index - 1;
```
why doesnt this case need a recursive call?

mcilroyc replied:
Since we're always starting at the "top", I believe this is safe.

pineapplespatula replied:
will we ever not start from the "top"? it seems like an easy fix.

mcilroyc replied:
I hear what you're saying, but 1) this passed all my tests and I'd rather not have to retest, because 2) this feature will probably have some tweaks after dave/seth get their hands on it anyway, so don't want to over do it just yet.

pineapplespatula commented:
@mcilroyc  just a few questions and nits. feel free to rebase

mcilroyc commented:
@pineapplespatula I addressed your comments in my latest commit and/or answered questions above.  I've rebased partially.  I this new level of granularity is what I want.  It is still a "lot of commits" but this is a big PR.  OK?

pineapplespatula commented:
I am curious why these commits weren't individual PRs.

pineapplespatula commented:
Good enough, with such a large PR i cant promise i didn't miss anything, but I'm going to go ahead and merge it now

mcilroyc commented:
rebased, fixed one conflict.  back to @pineapplespatula.

The reason this has so much in one PR: I did it over labor day weekend and there were a few times where fixes got intermingled amongst my various local branches, and THEN there were some rebases due to other open PRs.... and well, this was just safer.  won't happen again.  well, it probably will, but at least I'll try not to.
